lib/promscrape/discovery/puppetdb/resource.go

```diff
@@ -0,0 +1,170 @@
+package puppetdb
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/promscrape/discoveryutils"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/promutils"
+)
+
+const (
+	separator = ","
+)
+
+var matchContentTypeRegex = regexp.MustCompile(`^(?i:application\/json(;\s*charset=("utf-8"|utf-8))?)$`)
+
+type resource struct {
+	Certname    string     `json:"certname"`
+	Resource    string     `json:"resource"`
+	Type        string     `json:"type"`
+	Title       string     `json:"title"`
+	Exported    bool       `json:"exported"`
+	Tags        []string   `json:"tags"`
+	File        string     `json:"file"`
+	Environment string     `json:"environment"`
+	Parameters  parameters `json:"parameters"`
+}
+
+type parameters map[string]interface{}
+
+// toLabels convert Parameters map into label-value map.
+// See: https://github.com/prometheus/prometheus/blob/685493187ec5f5734777769f595cf8418d49900d/discovery/puppetdb/resources.go#L39
+func (p *parameters) toLabels() map[string]string {
+	if p == nil {
+		return nil
+	}
+	m := make(map[string]string, len(*p))
+
+	for k, v := range *p {
+		var labelValue string
+		switch value := v.(type) {
+		case string:
+			labelValue = value
+		case bool:
+			labelValue = strconv.FormatBool(value)
+		case int64:
+			labelValue = strconv.FormatInt(value, 10)
+		case float64:
+			labelValue = strconv.FormatFloat(value, 'g', -1, 64)
+		case []string:
+			labelValue = separator + strings.Join(value, separator) + separator
+		case []interface{}:
+			if len(value) == 0 {
+				continue
+			}
+			values := make([]string, len(value))
+			for i, v := range value {
+				switch value := v.(type) {
+				case string:
+					values[i] = value
+				case bool:
+					values[i] = strconv.FormatBool(value)
+				case int64:
+					values[i] = strconv.FormatInt(value, 10)
+				case float64:
+					values[i] = strconv.FormatFloat(value, 'g', -1, 64)
+				case []string:
+					values[i] = separator + strings.Join(value, separator) + separator
+				}
+			}
+			labelValue = strings.Join(values, separator)
+		case map[string]interface{}:
+			subParameter := parameters(value)
+			prefix := discoveryutils.SanitizeLabelName(k + "_")
+			for subk, subv := range subParameter.toLabels() {
+				m[prefix+subk] = subv
+			}
+		default:
+			continue
+		}
+		if labelValue == "" {
+			continue
+		}
+		name := discoveryutils.SanitizeLabelName(k)
+		m[name] = labelValue
+	}
+	return m
+}
+
+func getResourceList(cfg *apiConfig) ([]resource, error) {
+	body := struct {
+		Query string `json:"query"`
+	}{cfg.query}
+
+	bodyBytes, err := json.Marshal(body)
+	if err != nil {
+		return nil, err
+	}
+
+	modifyRequestFunc := func(request *http.Request) {
+		request.Body = io.NopCloser(strings.NewReader(string(bodyBytes)))
```

Comment:
I'd suggest to use bytes.Buffer instead:
`request.Body = io.NopCloser(bytes.NewBuffer(bodyBytes))`